File: 2020/day7.rb
# typed: strict
require 'sorbet-runtime'
require 'set'

extend T::Sig

class Graph
  extend T::Sig

  sig {void}
  def initialize()
    @nodes = T.let({}, T::Hash[String, Node])
  end

  sig {params(node_name: String).returns(Node)}
  def add_node(node_name)
    return nodes.fetch(node_name) if nodes.key?(node_name)
    nodes[node_name] = Node.new(node_name)
  end

  sig {params(node_name: String).returns(Node)}
  def find(node_name)
    nodes.fetch(node_name)
  end

  private

  sig{returns(T::Hash[String, Node])}
  attr_accessor :nodes

  class Node
    extend T::Sig

    sig{returns(String)}
    attr_reader :name

    class AnnotatedChild < T::Struct
      const :count, Integer
      const :node, Node
    end

    sig {params(name: String).void}
    def initialize(name)
      @name = name
      @children = T.let({}, T::Hash[String, AnnotatedChild])
      @parents = T.let(Set.new, T::Set[Node])
    end

    sig {params(node: Node).void}
    def add_parent(node)
      parents.add(node)
    end

    sig {params(node: Node, count: Integer).void}
    def add_child(node, count)
      raise if children.key?(node.name)
      node.add_parent(self)
      children[node.name] = AnnotatedChild.new(count: count, node: node)
    end

    sig {returns(T::Set[Node])}
    def get_all_parents
      Set.new.tap do |result|
        parents.each do |parent|
          result.add(parent)
          result.merge(parent.get_all_parents)
        end
      end
    end

    private

    sig{returns(T::Hash[String, AnnotatedChild])}
    attr_accessor :children

    sig {returns(T::Set[Node])}
    attr_accessor :parents
  end
end

sig {params(rules: T::Array[String]).returns(Graph)}
def parse_rules(rules)
  Graph.new.tap do |graph|
    rules.map do |rule|
      bag_type, contents = rule.gsub(/ bags?/, '').tr("\n.", '').split(" contain ")
      parent = graph.add_node(T.must(bag_type))

      T.must(contents).split(', ').each do |content|
        break if content == "no other"
        count, nested_bag_type = T.must(content.match(/(\d+) ([\w ]+)/)).captures
        parent.add_child(graph.add_node(T.must(nested_bag_type)), T.must(count).to_i)
      end
    end
  end
end

sig {params(graph: Graph).returns(Integer)}
def part_one(graph)
  graph.find("shiny gold").get_all_parents.count
end

if __FILE__ == $0
  rules = File.read("day7-input").lines
  graph = parse_rules(rules)
  puts "part 1 result: #{part_one(graph)}"
end

File: 2020/ts/day7.ts
import * as fs from 'fs';

class Node {
  name: string
  parents: Map<string, Node> = new Map()
  children: Map<string, Node> = new Map()

  constructor(name: string) {
    this.name = name
  }

  addParent(node: Node): void {
    this.parents.set(node.name, node)
  }

  addChild(node: Node): void {
    this.children.set(node.name, node)
    node.addParent(this)
  }

  getAllParents(): Set<Node> {
    let result: Set<Node> = new Set()
    this.parents.forEach(parent => {
      result.add(parent)
      result = new Set([...parent.getAllParents(), ...result])
    })
    return result
  }
}

class Graph {
  nodes: { [name: string]: Node } = {}

  addNode(name: string): Node {
    const existing = this.nodes[name]
    if (existing !== undefined) {
      return existing
    }
    return this.nodes[name] = new Node(name)
  }

  find(name: string): Node {
    const node = this.nodes[name]
    if (node === undefined) throw new Error("couldn't find that node")
    return node
  }
}

let graph = new Graph()

const lines = fs.readFileSync('../input/day7','utf8')
    .split("\n")

lines.forEach(line => {
  let [bagType, contents] = line.replaceAll(/ bags?/g, '').replaceAll(/[\n\.]/g, '').split(' contain ')
  if (bagType != undefined && contents != undefined) {
    let parent = graph.addNode(bagType)

    contents.split(', ').forEach(content => {
      let matches = content.match(/(?<count>\d+) (?<nested_bag_type>[\w ]+)/)?.groups
      if (matches?.nested_bag_type != undefined) {
        parent.addChild(graph.addNode(matches.nested_bag_type))
      }
    })
  }
})

console.log(graph.find("shiny gold").getAllParents().size)
